### PR TITLE
Inspect sub-folders recursively

### PR DESCRIPTION
**Pull request by doyou89**
This solves errors in walk_packages() when `is_pkg` is true.

### CODE REVIEW

**Comment by lucsorel** [comment]
thank you @doyou89 for your contribution.

Before your pull-request is merged, could you add a unit-test asserting that the generated plantuml doc matches is what one expects when recursively including a sub-folder and a sub-sub folder, if you please?

You can get some inspiration with the current unit tests.

**Comment by doyou89** [comment]
New test case for py2puml() method is added.
It uses "py2puml/domain" folder.
For testing sub-folder case, some folders need to be added in this folder.
But I cannot add examples in here.

File: py2puml/parser.py [hunk 1]
```diff
@@ -132,12 +132,15 @@ def parse_type(
 ):
     definition_type_fqdn = f'{definition_type.__module__}.{definition_type.__name__}'
     if definition_type_fqdn not in domain_items_by_fqdn:
-        if issubclass(definition_type, Enum):
-            parse_enum_type(definition_type, definition_type_fqdn, domain_items_by_fqdn)
-        elif getattr(definition_type, '_fields', None) is not None:
-            parse_namedtupled_class(definition_type, definition_type_fqdn, domain_items_by_fqdn)
-        else:
-            parse_class_type(definition_type, definition_type_fqdn, root_module_name, domain_items_by_fqdn, domain_relations)
+        try:
```
Comment (lucsorel) [inline py2puml/parser.py:135]: why do you need this try-catch block? Errors should not be silenced.

File: py2puml/py2puml.py [hunk 2]
```diff
@@ -1,7 +1,7 @@
 from typing import Dict, List, Iterable
 from types import ModuleType
-from pkgutil import walk_packages
-from importlib import import_module
+import pkgutil
+import importlib
```
Comment (lucsorel) [inline py2puml/py2puml.py:4]: please, do not use global imports. Can you revert these changes and follow the approach used throughout the codebase of this library, please? For consistency sake.

File: tests/py2puml/test_py2puml.py [hunk 3]
```diff
@@ -1,5 +1,47 @@
...
 
 
 def test_version():
-    assert __version__ == '0.3.0'
+    assert __version__ == '0.3.1'
+
+
+def test_py2puml():
+    """test py2puml on py2puml/domain."""
+    out = py2puml.py2puml('py2puml/domain', 'py2puml.domain')
```
Comment (lucsorel) [inline tests/py2puml/test_py2puml.py:10]: can you move this line below the expected output, so that the actual output and the assertion are close to each other (for readability sake), please?

And also, please rename the `out` variable into something more meaningful: `py2puml_domain` for example.

**Review by lucsorel — COMMENTED** [review]
I see. Yes, the folder with the py2puml domain should not be modified for testing purposes.

To add a proper unit test:

1. add a new domain script in the testing part of the codebase (https://github.com/lucsorel/py2puml/tree/master/tests/modules). Maybe something like `withsubdomain.py`
2. add a `subdomain` folder with a `subdomain.py` script within it, defining a subdomain Python class
3. make sure the  `withsubdomain.py` refers to the class in the `subdomain/subdomain.py` script
4. add a test in https://github.com/lucsorel/py2puml/blob/master/tests/py2puml/test_parser.py

**Comment by lucsorel** [comment]
also, feel free to add yourself in the [Contributions](https://github.com/lucsorel/py2puml/blob/master/README.md#contributions) section of the README.md file :smiley: 

```
* [your name as you want it to appear](https://github.com/doyou89)
```

File: py2puml/parser.py (hunk 1 above)
Comment (doyou89) [reply py2puml/parser.py:135]: Yes, I have misunderstanding for this code. I'll get back to original.

File: py2puml/py2puml.py (hunk 2 above)
Comment (doyou89) [reply py2puml/py2puml.py:4]: Ok I will. This change added for mocking easily in the unit test, but now unit test case have no mock object.

**Comment by doyou89** [comment]
Thank you. I added my name :smiley:

Unit test for subdomain added. please check the test meet to your idea.

File: py2puml/parser.py [hunk 4]
```diff
@@ -132,12 +132,13 @@ def parse_type(
...
-            parse_namedtupled_class(definition_type, definition_type_fqdn, domain_items_by_fqdn)
-        else:
-            parse_class_type(definition_type, definition_type_fqdn, root_module_name, domain_items_by_fqdn, domain_relations)
+        if isinstance(definition_type, type):   # process only class object
+            if issubclass(definition_type, Enum):
+                parse_enum_type(definition_type, definition_type_fqdn, domain_items_by_fqdn)
+            elif getattr(definition_type, '_fields', None) is not None:
+                parse_namedtupled_class(definition_type, definition_type_fqdn, domain_items_by_fqdn)
+            else:
+                parse_class_type(definition_type, definition_type_fqdn, root_module_name, domain_items_by_fqdn, domain_relations)
```
Comment (doyou89) [inline py2puml/parser.py:141]: Found why i changed this. if `definition_type` is not a class type, `issubclass(definition_type, Enum)` raise error.

File: tests/py2puml/test_py2puml.py (hunk 3 above)
Comment (doyou89) [reply tests/py2puml/test_py2puml.py]: Yes, i updated test case.

**Review by doyou89 — COMMENTED** [review]
If `definition_type` is not a class type, `issubclass(definition_type, Enum)` raise error.

**Comment by lucsorel** [comment]
thank you @doyou89 for your changes.

Today is a busy day for me, but I will have time to check your contribution tomorrow. Have a nice day!

**Comment by lucsorel** [comment]
Hi @doyou89 ,

Sorry to keep you waiting. I had a thorough look at the pull request and the propose changes do not implement a recursive exploration of the subfolders of the folder being processed. Moreover you added test classes in the domain model used by the library to generate the plantuml code. I therefore pushed 2 commits:

1. https://github.com/lucsorel/py2puml/pull/8/commits/293b1065c91b39725fee45cac2a9fb2232430b76: removal of the test files from the production code. I also removed the `if isinstance(definition_type, type):` condition which you had to add because [this import](https://github.com/lucsorel/py2puml/pull/8/commits/293b1065c91b39725fee45cac2a9fb2232430b76#diff-9bc47dfcedb2b6a94f71873f159709063ba2607ddec049693231a62c84bbd8f3L1) was invalid (the imported class didn't exist)
2. https://github.com/lucsorel/py2puml/pull/8/commits/15084dd59d4365b80e1e7826247d9ec61b9d9fc7: addition of 2 unit tests of the recursive exploration (which I prefer to be optional, for backward compatibility). Here is the situation handled by the test
```
/
┣tests
┃    ┗ modules
┃         ┗ withsubdomain
┃           ┣ withsubdomain.py: declares the Car class, which refers to the imported ...subdomain.Engine class
┃           ┗ subdomain
┃             ┗ insubdomain.py: declares the Engine class and the Pilot class
```

Expected behavior when the library explores the `tests/modules/withsubdomain` folder:
- when `recursive==False`: the plantuml diagram contains the Car class and the imported Engine class, but not the Pilot class because it is not visible from (not imported in) the `withsubdomain` folder. This is what the library actually does:



- when `recursive==True`: the plantuml diagram contains the Car class and the imported Engine class, and also the Pilot class because the subdomain folder has been also explored. This is what the library should produce:



Do you agree with these specifications? Is it also what you expected?

**Review by doyou89 — COMMENTED** [review]
`walk_packages` walks through sub-directories.
But each sub-directories should have `__init__.py` file to be recognized as python module.
I added this `__init__.py` file into modules and it's sub folders.
But in this case, `Pilot` cannot be excepted.

File: py2puml/parser.py [hunk 5]
```diff
@@ -132,13 +132,12 @@ def parse_type(
...
-            elif getattr(definition_type, '_fields', None) is not None:
-                parse_namedtupled_class(definition_type, definition_type_fqdn, domain_items_by_fqdn)
-            else:
-                parse_class_type(definition_type, definition_type_fqdn, root_module_name, domain_items_by_fqdn, domain_relations)
+        if issubclass(definition_type, Enum):
+            parse_enum_type(definition_type, definition_type_fqdn, domain_items_by_fqdn)
+        elif getattr(definition_type, '_fields', None) is not None:
+            parse_namedtupled_class(definition_type, definition_type_fqdn, domain_items_by_fqdn)
+        else:
+            parse_class_type(definition_type, definition_type_fqdn, root_module_name, domain_items_by_fqdn, domain_relations)
```
[the remark this answers is not in the document]
Comment (lucsorel) [reply py2puml/parser.py]: Thank you for your inputs! I have just realized that the initial code was already recursive :smile: 

I have pushed two commits:
- non-class imports must be filtered out at the proper place: https://github.com/lucsorel/py2puml/pull/8/commits/1a270aa6d6bb0383edd365be41013e3b0fa86d81. `parse_type` handles types (classes) and should not be given a function
- minor changes (project and poetry configurations, gitignore, upgrade of pytest): https://github.com/lucsorel/py2puml/pull/8/commits/816f91c96b16592fbc6504b3b904c871f907dc52

Please, never push [commented-out code like this](https://github.com/lucsorel/py2puml/pull/8/commits/1a270aa6d6bb0383edd365be41013e3b0fa86d81#diff-1f94fc1a3b623c6fbbeaa3b7d3b93c27883ec7aaa96abfbf2c9e93e7dbbba3fc). I believe you did it not to offend me (thanks :smile:) and it could be ok in a 'work-in-progress' commit, but not at the stage of a pull request.

If you agree with these commits, I am ready to merge your contribution. Then I can compile the library and push the new version on pypi.

[the remark this answers is not in the document]
Comment (doyou89) [reply py2puml/parser.py]: I totally agree with you. Thank you for great work!

[the remark this answers is not in the document]
Comment (doyou89) [reply py2puml/parser.py]: And one thing more. I added `__init__.py` in `tests` folder to run pytest in vscode.

**Comment by lucsorel** [comment]
@doyou89 version 0.3.1 is online and ready to be installed where you need it: https://pypi.org/project/py2puml/. Thank you :smiley:
